AX: Compute AXAttributedStringForTextMarkerRangeAttribute off the main-thread for ranges pointing to static text objects

twilco opened the pull request:
#### cc3d1eef8dbf47a550477853ab61012aa91192e9
<pre>
AX: Compute AXAttributedStringForTextMarkerRangeAttribute off the main-thread for ranges pointing to static text objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=283607">https://bugs.webkit.org/show_bug.cgi?id=283607</a>
<a href="https://rdar.apple.com/problem/140447302">rdar://problem/140447302</a>

Reviewed by Chris Fleizach.

This commit implements off-main-thread support for AXAttributedStringForTextMarkerRangeAttribute for text marker ranges
that point within a single static text object. Supporting ranges that span multiple objects, and ranges that are confined
to one object that is not static text, will be implemented in future patches. These both should be much easier with this
base laid down, as fundamentally they will need to simply move from static text to static text, using the code written
in this commit to build a combined attributed string.

Beyond setting style attributes on attributed strings, there are several other categories of attributes that this commit
leaves for a future patch, e.g. heading level, blockquote level, expanded text, and a few more.

Several new isolated tree properties were added to support this functionality. Once this work is complete and ships, several
existing stale cache bugs will be fixed (we do not currently update cached attributed strings when style changes).

With this commit, we now pass attributed-string/attributed-string-text-styling.html with AX_THREAD_TEXT_APIS_ENABLED.
A new test is also created to ensure that dynamic style changes are reflected in an object&apos;s computed attributed string:
ax-thread-text-apis/attributed-string/dynamic-style.html

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/dynamic-style-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/dynamic-style.html: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::LineDecorationStyle::LineDecorationStyle):
(WebCore::LineDecorationStyle::debugDescription const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::LineDecorationStyle::LineDecorationStyle):
(WebCore::AttributedStringStyle::hasUnderline const):
(WebCore::AttributedStringStyle::underlineColor const):
(WebCore::AttributedStringStyle::hasLinethrough const):
(WebCore::AttributedStringStyle::linethroughColor const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::lineNumberForIndex const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm: Added.
(WebCore::attributedStringContainsRange):
(WebCore::attributedStringSetFont):
(WebCore::attributedStringSetColor):
(WebCore::attributedStringSetStyle):
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::fontFrom):
(WebCore::textColorFrom):
(WebCore::backgroundColorFrom):
(WebCore::AccessibilityObject::font const):
(WebCore::AccessibilityObject::textColor const):
(WebCore::AccessibilityObject::backgroundColor const):
(WebCore::AccessibilityObject::isSubscript const):
(WebCore::AccessibilityObject::isSuperscript const):
(WebCore::AccessibilityObject::hasTextShadow const):
(WebCore::AccessibilityObject::lineDecorationStyle const):
(WebCore::AccessibilityObject::stylesForAttributedString const):
(WebCore::attributedStringContainsRange): Deleted.
(WebCore::attributedStringSetFont): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::updateNodeProperty):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
(WebCore::AXIsolatedObject::stylesForAttributedString const):
(WebCore::AXIsolatedObject::textMarkerRange const):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::attributedStringSetColor): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::styleDidChange):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: <a href="https://commits.webkit.org/287243@main">https://commits.webkit.org/287243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18f06cdb4766a34498f30675b0a0043d8bf04420

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77904 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82552 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29161 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80025 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66097 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60999 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18934 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50989 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66909 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41302 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24366 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27504 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24719 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83914 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5271 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3550 "Found 2 new test failures: ipc/create-media-source-with-invalid-constraints-crash.html webrtc/vp8-then-h264.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69220 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68476 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17252 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12473 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10593 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5219 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7972 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5211 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6996 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
```diff
@@ -0,0 +1,175 @@
+        [fontAttributes setValue:bridge_cast(fullName.get()) forKey:AccessibilityTokenFontName];
+    if (familyName)
+        [fontAttributes setValue:bridge_cast(familyName.get()) forKey:AccessibilityTokenFontFamily];
+    if ([size boolValue])
```
this is a NSNumber wrapping a float. So boolValue might not be correct

twilco replied:
I think it's checking whether the `NSNumber` is greater than zero (but maybe `boolValue` doesn't work this way, I'm not sure) 🤔 and if so, we set the font size with the actual NSNumber value:

```objc
if ([size boolValue])
    [fontAttributes setValue:size forKey:AccessibilityTokenFontSize];
```

But I just moved this function from elsewhere, didn't change it at all. If there's a better way to do this I'm happy to change it though, just let me know what you think we should do!

fleizach commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3609,8 +3609,18 @@ - (id)accessibilityAttributeValue:(NSString*)attribute forParameter:(id)paramete
+#if ENABLE(AX_THREAD_TEXT_APIS)
+        // FIXME: Expand this beyond static text (i.e. ranges that span multiple objects).
+        if (AXObjectCache::useAXThreadTextApis() && backingObject->isStaticText()) {
+            if (!textMarkerRange)
```
this could be moved outside the ENABLE(AX_THREAD) block

twilco replied:
Fixed in latest commit!

twilco committed:
AX: Compute AXAttributedStringForTextMarkerRangeAttribute off the main-thread for ranges pointing to static text objects
https://bugs.webkit.org/show_bug.cgi?id=283607
rdar://problem/140447302

Reviewed by Chris Fleizach.

This commit implements off-main-thread support for AXAttributedStringForTextMarkerRangeAttribute for text marker ranges
that point within a single static text object. Supporting ranges that span multiple objects, and ranges that are confined
to one object that is not static text, will be implemented in future patches. These both should be much easier with this
base laid down, as fundamentally they will need to simply move from static text to static text, using the code written
in this commit to build a combined attributed string.

Beyond setting style attributes on attributed strings, there are several other categories of attributes that this commit
leaves for a future patch, e.g. heading level, blockquote level, expanded text, and a few more.

Several new isolated tree properties were added to support this functionality. Once this work is complete and ships, several
existing stale cache bugs will be fixed (we do not currently update cached attributed strings when style changes).

With this commit, we now pass attributed-string/attributed-string-text-styling.html with AX_THREAD_TEXT_APIS_ENABLED.
A new test is also created to ensure that dynamic style changes are reflected in an object's computed attributed string:
ax-thread-text-apis/attributed-string/dynamic-style.html

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/dynamic-style-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/dynamic-style.html: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::LineDecorationStyle::LineDecorationStyle):
(WebCore::LineDecorationStyle::debugDescription const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::LineDecorationStyle::LineDecorationStyle):
(WebCore::AttributedStringStyle::hasUnderline const):
(WebCore::AttributedStringStyle::underlineColor const):
(WebCore::AttributedStringStyle::hasLinethrough const):
(WebCore::AttributedStringStyle::linethroughColor const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::characterRangeForLine const):
(WebCore::AXTextMarker::lineNumberForIndex const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm: Added.
(WebCore::attributedStringContainsRange):
(WebCore::attributedStringSetFont):
(WebCore::attributedStringSetColor):
(WebCore::attributedStringSetStyle):
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::fontFrom):
(WebCore::textColorFrom):
(WebCore::backgroundColorFrom):
(WebCore::AccessibilityObject::font const):
(WebCore::AccessibilityObject::textColor const):
(WebCore::AccessibilityObject::backgroundColor const):
(WebCore::AccessibilityObject::isSubscript const):
(WebCore::AccessibilityObject::isSuperscript const):
(WebCore::AccessibilityObject::hasTextShadow const):
(WebCore::AccessibilityObject::lineDecorationStyle const):
(WebCore::AccessibilityObject::stylesForAttributedString const):
(WebCore::attributedStringContainsRange): Deleted.
(WebCore::attributedStringSetFont): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::updateNodeProperty):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
(WebCore::AXIsolatedObject::stylesForAttributedString const):
(WebCore::AXIsolatedObject::textMarkerRange const):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::attributedStringSetColor): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::styleDidChange):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveComposedTree):

Canonical link: https://commits.webkit.org/287243@main
